pkg/agent/cniserver/pod_configuration.go

```diff
@@ -251,26 +249,33 @@ func configureInterface(
 		}
 	}()
 
-	// Configure IP for container
-	klog.V(2).Infof("Configuring IP address for container %s", containerID)
-	if err = configureContainerAddr(netns, containerIface, result); err != nil {
-		klog.Errorf("Failed to configure IP address on container %s:%v", containerID, err)
-		return fmt.Errorf("failed to configure container ip")
-	}
-
 	// GetOFPort will wait for up to 1 second for OVSDB to report the OFPort number.
 	ofPort, err := ovsBridge.GetOFPort(ovsPortName)
 	if err != nil {
 		klog.Errorf("Failed to get of_port of OVS interface %s: %v", ovsPortName, err)
 		return err
 	}
-	// Setup openflow entries for OVS interface
-	klog.V(2).Infof("Setting up openflow entries for container %s", containerID)
+	klog.V(2).Infof("Setting up Openflow entries for container %s", containerID)
 	err = ofClient.InstallPodFlows(ovsPortName, containerConfig.IP, containerConfig.MAC, gateway.MAC, uint32(ofPort))
 	if err != nil {
-		klog.Errorf("Failed to add openflow entries for container %s: %v", containerID, err)
+		klog.Errorf("Failed to add Openflow entries for container %s: %v", containerID, err)
 		return err
 	}
+	defer func() {
+		if !success {
+			ofClient.UninstallPodFlows(ovsPortName)
```

Comment:
My only question is - if it is easy to clean up flows. I think now we do not do periodic GC and the flow might be left there for ever, if UninstallPodFlows fails?

Should we consider to move GARP sending out of configureContainerAddr, and still do configureContainerAddr before InstallPodFlows?

Reply:
Normally the last step `configureContainerAddr` shouldn't fail as it's not related to OVS userspace, unless the netns became invalid. Even if GARP is moved out of `configureContainerAddr`, sending GARP also needs to enter that netns and look for the target interface again and might encounter the same error (if netns is invalid), then we still need to try to rollback the PodFlows and have the same problem, which looks like not helpful to address the stale flows.
In theory, the case only happens when there are something wrong with OVS and container netns at the same time, and if it really happens, we would have orphan OVS port and flow, but they look like not harmful to the traffic in worst case and would be removed upon restarting. So it should be not very bad?

Reply:
My thinking is no need to rollback for GARP sending errors. It is just a try to refresh ARP caches for potential conflicts.

Reply:
Before the change, the code have ignored the error of sending GARP. To send GARP in a separate function, it will need to enter ns and look for interface too, the possible errors are missing netns and interface, do we want to ignore them? If not, it seems no difference from sending GARP in configureContainerAddr, which could save entering netns and looking for interface twice.

Reply:
Thought over. I am fine with your current change. Anyway, UninstallPodFlows() could still fail in Pod deletion, and we still need a way to clean up flows.